uber/site_sections/preregistration.py

```diff
@@ -524,3 +527,22 @@ def shirt_reorder(self, session, message='', **params):
             'message': message,
             'attendee': attendee
         }
+
+    def apply_promo_code(self, session, message='', **params):
+        if 'id' and 'code' in params:
+            if params['code'] is not '':
+                attendee, group = self._get_unsaved(params['id'])
+                if not attendee:
+                    message = 'Attendee Not Found'
+                else:
+                    promo_code = session.query(PromoCode).filter(PromoCode.code == params['code'] and not PromoCode.expired).first()
```

Comment:
Wait, does this actually work? I didn't think you could check a property using the SQLAlchemy ORM.

In any case, you're doing two checks here but only returning an error message for one. It'd be more user-friendly to fetch the promo code first, then return a DIFFERENT error if the promo code doesn't exist vs if it's all used up vs if the expiration date has passed.